mlflow/models/evaluation/__init__.py

```diff
@@ -0,0 +1,399 @@
+        """
+        :param data: One of the following:
+         - A numpy array or list of evaluation features, excluding labels.
+         - A Pandas DataFrame, or the path to a serialized DataFrame,
+           containing evaluation features and labels. All columns will be regarded as feature
+           columns except the "labels" column.
+
+        :param labels: One of the following:
+         - A numpy array or list of evaluation labels, if `data` is also a numpy array or list.
+         - The string name of a column from `data` that contains evaluation labels, if `data`
+           is a DataFrame.
+
+        :param name: (Optional) The name of the dataset (must not contain ").
+
+        :param path: (Optional) the path to a serialized DataFrame (must not contain ").
+          (e.g. a delta table, parquet file)
+        """
+        if name is not None and '"' in name:
+            raise ValueError(f'Dataset name cannot include " but got name {name}')
+        if path is not None and '"' in path:
+            raise ValueError(f'Dataset path cannot include " but got name {path}')
+
+        if isinstance(data, (np.ndarray, list)):
+            if not isinstance(labels, (np.ndarray, list)):
+                raise ValueError(
+                    'If data is a numpy array or list of evaluation features, '
+                    'labels must be a numpy array or list of evaluation labels'
+                )
+        elif isinstance(data, pd.DataFrame):
+            if not isinstance(labels, str):
+                raise ValueError(
+                    'If data is a Pandas DataFrame, labels must be the string name of a column '
+                    'from `data` that contains evaluation labels'
+                )
+        else:
+            raise ValueError('The data argument must be a numpy array, a list or a '
+                             'Pandas DataFrame.')
+
+        self._user_specified_name = name
+        self.data = data
+        self.labels = labels
+        self.path = path
+        self._hash = None
+
+    def _extract_features_and_labels(self):
+        if isinstance(self.data, np.ndarray):
+            return self.data, self.labels
+        elif isinstance(self.data, pd.DataFrame):
+            feature_cols = [x for x in self.data.columns if x != self.labels]
+            return self.data[feature_cols], self.data[self.labels]
+        else:
+            raise ValueError(f'Unsupported data type: {type(self.data)}')
+
+    @staticmethod
+    def _array_like_obj_to_bytes(data):
+        if isinstance(data, pd.DataFrame):
+            return data.to_numpy().tobytes()
+        elif isinstance(data, np.ndarray):
+            return data.tobytes()
+        elif isinstance(data, list):
+            return np.array(data).tobytes()
+        else:
+            raise ValueError('Unsupported data type.')
+
+    @staticmethod
+    def _gen_md5_for_arraylike_obj(md5_gen, data):
+        md5_gen.update(pickle.dumps(len(data)))
```

Comment:
Do we need to call `pickle.dumps()` on `len(data)`, or can we just pass in the len since it's an int?

Reply:
changed to `md5_gen.update(np.int64(len(data)).tobytes())` to avoid calling pickle here